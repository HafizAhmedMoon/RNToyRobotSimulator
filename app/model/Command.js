export class Command {

  static COMMAND_DICT = {
    PLACE: 'PLACE',
    LEFT: 'LEFT',
    RIGHT: 'RIGHT',
    MOVE: 'MOVE',
    REPORT: 'REPORT',
  };

  static parse(command) {
    const [cmd, ...args] = command.trim().split(' ');
    return new Command(cmd, args);
  }

  command = '';
  args = [];

  constructor(command, args) {
    this.command = command;
    this.args = args;
  }

  get isInvalid() {
    return !Command.COMMAND_DICT[this.command]
  }
}
