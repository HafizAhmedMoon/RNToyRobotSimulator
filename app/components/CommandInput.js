import PropTypes from 'prop-types';
import React from 'react';
import { Command } from "../model/Command";
import { StyleSheet, TextInput, View } from "react-native";

export class CommandInput extends React.Component {

  static propTypes = {
    onCommandReceive: PropTypes.func.isRequired
  };

  state = {
    command: ''
  };

  onSubmit() {
    const {onCommandReceive} = this.props;
    const {command} = this.state;
    onCommandReceive(Command.parse(command));

    this.setState({command: ''});
    this.refs.commandInput.clear();
  }

  render() {
    return (
      <View style={styles.container}>
        <TextInput
          ref="commandInput"
          onSubmitEditing={this.onSubmit.bind(this)}
          placeholder="Input Command"
          keyboardType="default"
          returnKeyType="done"
          style={styles.input}
          onChangeText={command => this.setState({command})}
          disableFullscreenUI={true}

          underlineColorAndroid="transparent"
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignSelf: 'stretch',
    paddingHorizontal: 10
  },
  input: {
    borderColor: '#ddd',
    borderWidth: 0,
    borderBottomWidth: 1,
    paddingBottom: 0
  }
});
