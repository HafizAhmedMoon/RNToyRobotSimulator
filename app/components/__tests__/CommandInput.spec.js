import React from 'react';
import { shallow } from 'enzyme';
import { CommandInput } from '../CommandInput';
import { Command } from "../../model/Command";

describe("CommandInput Component", () => {
  it('renders as expected', () => {
    const onCommandReceive = jest.fn();
    const wrapper = shallow(
      <CommandInput onCommandReceive={onCommandReceive}/>
    );
    expect(wrapper).toMatchSnapshot();

    const TextInput = wrapper.find('TextInput');
    expect(TextInput).toHaveLength(1);

    TextInput.simulate('changeText', 'MOVE');
    expect(wrapper.state('command')).toBe('MOVE');

    try {
      TextInput.simulate('submitEditing'); // TODO: create proper TextInput mock to fix
    } catch (e) {
    }
    expect(onCommandReceive.mock.calls).toHaveLength(1);
    expect(onCommandReceive.mock.calls[0][0]).toBeInstanceOf(Command);
    expect(onCommandReceive.mock.calls[0][0]).toEqual(Command.parse('MOVE'));

    expect(wrapper.state('command')).toBe('');
  });
});
