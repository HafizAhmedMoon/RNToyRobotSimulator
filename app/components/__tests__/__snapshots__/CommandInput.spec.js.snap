// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommandInput Component renders as expected 1`] = `
<View
  style={
    Object {
      "alignSelf": "stretch",
      "paddingHorizontal": 10,
    }
  }
>
  <TextInput
    autoCapitalize="characters"
    disableFullscreenUI={true}
    keyboardType="default"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Input Command"
    returnKeyType="done"
    style={
      Object {
        "borderBottomWidth": 1,
        "borderColor": "#ddd",
        "borderWidth": 0,
        "paddingBottom": 0,
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
